ebean-core/src/main/java/io/ebeaninternal/server/persist/DefaultPersister.java

```diff
@@ -1123,7 +1125,7 @@ private void saveAssocOne(PersistRequestBean<?> request) {
           && !prop.isReference(detailBean)
           && !request.isParent(detailBean)) {
           SpiTransaction t = request.transaction();
-          t.depth(-1);
+          t.depthDecrement();
```

Comment:
The magic here - effectively we only decrement the depth when it is already greater than 0. This has the impact that it now will not make the depth negative (and hence order it before other persists requests at depth zero).

When we want to do this we now need to ensure we reset the depth back to zero after top level inserts and updates occur.

Reply:
Q: Is it OK to increment the depth two lines below, if we did not decrement it?
```java
int decremented = t.depthDecrement(); // returns 0 or 1
saveRecurse(...)
t.depth(decremented )
```
May also other places where depth is modified be affected by this kind of bug (e.g. `deleteAssocMany`, `deleteList` ...)
t.depth(decremented);

Reply:
> Q: Is it OK to increment the depth two lines below, if we did not decrement it?

Yes I think so.  I actually started off with code like what you have (I called decremented delta) but actually that just left things at depth 0 and was ok but I felt it was better to instead ~ "move everything down 1" in terms of depth.

>  deleteAssocMany, deleteList ...

I am not sure there is the same ordering issue with deletes yet. I'd prefer to have a failing test first.